### PR TITLE
improve test x/gamm

### PR DESCRIPTION
**Pull request by DongLieu**
Closes: #3849

## What is the purpose of the change

add corner cases and test cases as described

## Testing and Verifying

run test and lint

### CODE REVIEW

**Comment by faddat** [comment]
@DongLieu is there an issue with the description?  code looks good to me, but we should link to the issue if there is one :)

File: x/gamm/keeper/pool_service_test.go [hunk 1]
```diff
@@ -840,29 +842,46 @@ func (suite *KeeperTestSuite) TestJoinSwapExactAmountInConsistency() {
 		shareOutMinAmount sdk.Int
 		expectedSharesOut sdk.Int
 		tokenOutMinAmount sdk.Int
+		expectError       error
 	}{
 		{
-			name:              "single coin with zero swap and exit fees",
+			name:              "happy path: single coin with zero swap and exit fees",
 			poolSwapFee:       sdk.ZeroDec(),
 			poolExitFee:       sdk.ZeroDec(),
```
Comment (pysel) [inline x/gamm/keeper/pool_service_test.go:850]: I might be wrong, but as far as I am concerned, the exit fee was removed from osmosis pools via governance recently. I do not know if the governance's decision was already reflected on the protocol, but if yes, I think we should remove the `poolExitFee` field at all here

File: x/gamm/keeper/pool_service_test.go [hunk 2]
```diff
@@ -840,29 +842,46 @@ func (suite *KeeperTestSuite) TestJoinSwapExactAmountInConsistency() {
...
-		// 	expectedSharesOut: sdk.NewInt(6226484702880621000),
-		// 	tokenOutMinAmount: sdk.ZeroInt(),
-		// },
+		{
+			name:              "corner case: single coin with positive swap fee and zero exit fee",
+			poolSwapFee:       sdk.NewDecWithPrec(1, 2),
+			poolExitFee:       sdk.ZeroDec(),
+			tokensIn:          sdk.NewCoins(sdk.NewCoin("foo", sdk.NewInt(1_000_000))),
+			shareOutMinAmount: sdk.ZeroInt(),
+			expectedSharesOut: sdk.NewInt(6226484702880621000),
```
Comment (pysel) [inline x/gamm/keeper/pool_service_test.go:862]: I think it would be nice if we had a link to the calculations, which led to this number. Reference: https://github.com/osmosis-labs/osmosis/blob/main/osmomath/decimal_test.go#L724

Comment (pysel) [reply x/gamm/keeper/pool_service_test.go:862]: It might be out of scope of this PR, but I think it is not a big deal if we add it in it, if people agree with the importance of it

File: x/gamm/keeper/pool_service_test.go [hunk 3]
```diff
@@ -840,29 +842,46 @@ func (suite *KeeperTestSuite) TestJoinSwapExactAmountInConsistency() {
 		shareOutMinAmount sdk.Int
 		expectedSharesOut sdk.Int
 		tokenOutMinAmount sdk.Int
+		expectError       error
```
Comment (pysel) [inline x/gamm/keeper/pool_service_test.go:845]:
```suggestion
		expectedError       error
```

File: x/gamm/keeper/pool_service_test.go [hunk 4]
```diff
@@ -840,29 +842,46 @@ func (suite *KeeperTestSuite) TestJoinSwapExactAmountInConsistency() {
...
+			name:              "corner case: single coin with positive swap fee and zero exit fee",
+			poolSwapFee:       sdk.NewDecWithPrec(1, 2),
+			poolExitFee:       sdk.ZeroDec(),
+			tokensIn:          sdk.NewCoins(sdk.NewCoin("foo", sdk.NewInt(1_000_000))),
+			shareOutMinAmount: sdk.ZeroInt(),
+			expectedSharesOut: sdk.NewInt(6226484702880621000),
+			tokenOutMinAmount: sdk.ZeroInt(),
+		},
+		{
+			name:              "error: calculated amount is lesser than min amount",
```
Comment (pysel) [inline x/gamm/keeper/pool_service_test.go:866]: nit:
```suggestion
			name:              "error: calculated amount is less than min amount",
```

File: x/gamm/keeper/pool_service_test.go [hunk 5]
```diff
@@ -840,29 +842,46 @@ func (suite *KeeperTestSuite) TestJoinSwapExactAmountInConsistency() {
...
+		},
+		{
+			name:              "error: non zero exit fee",
+			poolSwapFee:       sdk.NewDecWithPrec(1, 2),
+			poolExitFee:       sdk.NewDecWithPrec(1, 2),
+			tokensIn:          sdk.NewCoins(sdk.NewCoin("foo", sdk.NewInt(1_000_000))),
+			shareOutMinAmount: sdk.ZeroInt(),
+			expectedSharesOut: sdk.NewInt(6265857020099440400),
+			tokenOutMinAmount: sdk.ZeroInt(),
+			expectError:       fmt.Errorf("can not create pool with non zero exit fee, got %v", sdk.NewDecWithPrec(1, 2)),
```
Comment (pysel) [inline x/gamm/keeper/pool_service_test.go:883]: nit:
```suggestion
			expectError:       fmt.Errorf("cannot create pool with non zero exit fee, got %v", sdk.NewDecWithPrec(1, 2)),
```

File: x/gamm/keeper/pool_service_test.go [hunk 6]
```diff
@@ -891,32 +910,34 @@ func (suite *KeeperTestSuite) TestJoinSwapExactAmountInConsistency() {
 					ExitFee: tc.poolExitFee,
 				},
 			)
+			if err != nil {
+				suite.Require().Equal(err, tc.expectError)
```
Comment (pysel) [inline x/gamm/keeper/pool_service_test.go:914]: I personally think that the expected error in unit tests should reflect an error that might happen inside a function we are testing, which is, in this case, `JoinSwapExactAmountIn`. I think if we want to test edge cases for balancer pool creation, it should be done in a separate test

Comment (pysel) [reply x/gamm/keeper/pool_service_test.go:914]: if, for any reason, something breaks for pool creation, when the expected error is related to `JoinSwapExactAmountIn`, this require statement will compare an error happened during pool creation with some internal to `JoinSwapExactAmountIn` edge case

File: x/gamm/keeper/pool_service_test.go [hunk 7]
```diff
@@ -891,32 +910,34 @@ func (suite *KeeperTestSuite) TestJoinSwapExactAmountInConsistency() {
...
+					suite.Require().True(tokenOutAmt.LTE(swapFeeAdjustedAmount))
+				}
+			}
 
-			// require swapTokenOutAmt + 10 > input
-			suite.Require().True(
-				swapFeeAdjustedAmount.Sub(tokenOutAmt).LTE(sdk.NewInt(10)),
-				"expected out amount %s, actual out amount %s",
-				swapFeeAdjustedAmount, tokenOutAmt,
-			)
```
Comment (pysel) [inline x/gamm/keeper/pool_service_test.go:919]: why are we removing this?

File: x/gamm/pool-models/balancer/pool_suite_test.go [hunk 8]
```diff
@@ -1095,4 +1095,36 @@ func (suite *KeeperTestSuite) TestRandomizedJoinPoolExitPoolInvariants() {
 	for i := 0; i < 50000; i++ {
 		testPoolInvariants()
 	}
+	testCaseCorners := []testCase{
+		{ //join pool for values greater than in the pool
```
Comment (pysel) [inline x/gamm/pool-models/balancer/pool_suite_test.go:1099]: I think this file is not `gofmt`'ed

File: x/gamm/pool-models/balancer/pool_suite_test.go [hunk 9]
```diff
@@ -1095,4 +1095,36 @@ func (suite *KeeperTestSuite) TestRandomizedJoinPoolExitPoolInvariants() {
...
+			initialTokensDenomIn:  1,
+			initialTokensDenomOut: 9_223_372_036_854_775_807,
+			percentRatio:          101,
+		},
+		{ //with percentRatio = 0
+			initialTokensDenomIn:  1,
+			initialTokensDenomOut: 9_223_372_036_854_775_807,
+			percentRatio:          0,
+		},
+		{ //
```
Comment (pysel) [inline x/gamm/pool-models/balancer/pool_suite_test.go:1109]:
```suggestion
		{
```

File: x/gamm/pool-models/balancer/pool_suite_test.go [hunk 10]
```diff
@@ -1095,4 +1095,36 @@ func (suite *KeeperTestSuite) TestRandomizedJoinPoolExitPoolInvariants() {
...
+			initialTokensDenomIn:  9_223_372_036_854_775_807,
+			initialTokensDenomOut: 1,
+			percentRatio:          101,
+		},
+		{
+			initialTokensDenomIn:  9_223_372_036_854_775_807,
```
Comment (pysel) [inline x/gamm/pool-models/balancer/pool_suite_test.go:1115]: asking to learn: is there a specific reason for this number? or does it just need to be big enough? if latter, big enough compared to what?

**Review by pysel — COMMENTED** [review]
Thanks for the great work! Left some comments:

File: x/gamm/keeper/pool_service_test.go (hunk 1 above)
Comment (DongLieu) [reply x/gamm/keeper/pool_service_test.go:850]: I tried changing it but it doesn't seem to work, I don't think it's necessary to change it yet

File: x/gamm/keeper/pool_service_test.go (hunk 5 above)
Comment (DongLieu) [reply x/gamm/keeper/pool_service_test.go:883]: Thank you for checking

File: x/gamm/pool-models/balancer/pool_suite_test.go (hunk 10 above)
Comment (DongLieu) [reply x/gamm/pool-models/balancer/pool_suite_test.go:1115]: it is the maximum number of type int64, I think the edge cases are the ones that lie at border

**Review by faddat — APPROVED** [review]

**Comment by faddat** [comment]
this looks good @DongLieu, thanks!  I'm just going to merge main and make sure it passes golangci-lint with:

```yaml
run:
  tests: true
```

**Review by faddat — CHANGES_REQUESTED** [review]
I'm pushing a commit that changes the linter to lint the tests, and updates to main.  Can you please handle the linting @DongLieu ?

Thanks!

**Review by faddat — APPROVED** [review]

**Comment by DongLieu** [comment]
Ok, thank sir!

**Comment by faddat** [comment]
@DongLieu -- theres a large block of comments that I've removed from pull_test.go -- nice work here!

**Comment by faddat** [comment]
@DongLieu -- I tried to merge this, it didn't work.  I'll revert my last commit.  Can you get main merged please?

**Comment by faddat** [comment]
Sir, I made several flubs here, and am force pushing to a recent commit of yours...

File: x/gamm/keeper/keeper_test.go [hunk 11]
```diff
@@ -39,16 +39,14 @@ func (suite *KeeperTestSuite) prepareCustomBalancerPool(
...
 	suite.fundAllAccountsWith(balances)
 
 	poolID, err := suite.App.PoolManagerKeeper.CreatePool(
 		suite.Ctx,
 		balancer.NewMsgCreateBalancerPool(suite.TestAccs[0], poolParams, poolAssets, ""),
 	)
-	suite.Require().NoError(err)
-
-	return poolID
+	return poolID, err
```
Comment (hieuvubk) [inline x/gamm/keeper/keeper_test.go:49]: Why do we need to bring `suite.Require().NoError(err)` to outside of this func for any purpose?

**Comment by hieuvubk** [comment]
There are many conflicts here @DongLieu . Could u resolve them?

File: x/gamm/keeper/keeper_test.go (hunk 11 above)
Comment (DongLieu) [reply x/gamm/keeper/keeper_test.go:49]: In issue #3849 there is suggest with corner cases 0 <= exitFee < 1
however cannot create pool with exitFee > 0. So i want to get that error outside

Above pysel says that "I could be wrong, but as far as I know, the exit fee was recently removed from the osmosis pool through the administration. I don't know if the admin's decision has been reflected. reflected in the protocol or not, but if so, I think we should delete this field poolExitFee here".
And I removed testcase poolExitFee > 0

**Comment by faddat** [comment]
thanks a lot!
